### PR TITLE
Add SecFilter, refactor Variable and Variables slightly to accomodate.

### PR DESCRIPTION
**Pull request by sovdeeth**
### Description

Adds a filter section to allow filtering on many conditions at once, while maintaining indices and adding significant performance optimization when only removing a few elements of a large list.

Also support `any` and `all`, like multiline ifs.

```
set {_a::*} to integers between -10 and 10

filter {_a::*} to match:
	input is a number
	mod(input, 2) = 0
	input > 0
```

I would like feedback on the abuses of the Variable api I have committed, as I don't really think the code is that great. I couldn't think of any better approaches, however.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** #6866

### CODE REVIEW

**Comment by Mwexim** [comment]
Wouldn't it be smart to also allow a lambda-like structure where one can return a boolean?

```
filter {_players::*}:
    if health of player > 10:
        return true
    return false

```

Not sure if Skript's current section API facilitates this feature, but it allows for more flexibility and is more intuitive in my opinion.

Either way, I like both options actually since your proposition is a really clean way to filter expressions without the cluttering '[]' syntax that is used currently.

**Comment by Fusezion** [comment]
Sections have access to a return api iirc in 2.9+ so this should be possible

**Comment by sovdeeth** [comment]
> Wouldn't it be smart to also allow a lambda-like structure where one can return a boolean?
> 
> ```
> filter {_players::*}:
>     if health of player > 10:
>         return true
>     return false
> ```
> 
> Not sure if Skript's current section API facilitates this feature, but it allows for more flexibility and is more intuitive in my opinion.
> 
> Either way, I like both options actually since your proposition is a really clean way to filter expressions without the cluttering '[]' syntax that is used currently.

I considered this, but I thought it could get messy, and at that point there's no benefit to having a filter section over just doing
```
loop {list::*}:
    if loop-value is x:
        delete {list::%loop-index%}
```

File: src/main/java/ch/njol/skript/sections/SecFilter.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		String varName = unfilteredObjects.getName().toString(event);
+		String varSubName = StringUtils.substring(varName, 0, -1);
+		boolean local = unfilteredObjects.isLocal();
+
+		// not ideal to get this AND the iterator, but using this value could be unreliable due to name change issue from above.
+		// since we just use it for a length optimization at the end, it's ok to be a little unreliable.
+		var rawVariable = ((Map<String, Object>) unfilteredObjects.getRaw(event));
+		if (rawVariable == null)
+			return getNext();
+		var initialSize = rawVariable.size();
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/sections/SecFilter.java:129]:
```suggestion
		int initialSize = rawVariable.size();
```

**Review by UnderscoreTud — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]

**Comment by TheAbsolutionism** [comment]
I have a suggestion question for this,
1) What about an "alias" for `input` called `value`
2) Matching the `value` if the list that is being looped, is a list variable, to also be able to use `index` of sorts. That way you can filter out the values and indices within one section, if a user wanted to

I know for `2` you could just simply do two filters, one that processes the values, then once that is done, to do the indices.
Was just curious though

**Comment by sovdeeth** [comment]
> I have a suggestion question for this,
> 
>     1. What about an "alias" for `input` called `value`
> 
>     2. Matching the `value` if the list that is being looped, is a list variable, to also be able to use `index` of sorts. That way you can filter out the values and indices within one section, if a user wanted to
> 
> 
> Example
> 
> ```
> filter {_a::*}:
>     value is an integer
>     index parsed as number is set
> ```
> 
> I know for `2` you could just simply do two filters, one that processes the values, then once that is done, to do the indices. Was just curious though

You can already use `input index`! I'm not sure if using `value` would cause any issues but it seems reasonable.

**Comment by TheAbsolutionism** [comment]
> You can already use `input index`! I'm not sure if using `value` would cause any issues but it seems reasonable.

Ahhh ok, I was unaware of the `input index`, my apologies

File: src/main/java/ch/njol/skript/variables/Variables.java [hunk 2]
```diff
@@ -474,6 +478,67 @@ public static Object getVariable(String name, @Nullable Event event, boolean loc
...
+			private Object next = null;
+
+			@Override
+			public boolean hasNext() {
+				if (next != null)
+					return true;
+				while (keys.hasNext()) {
+					key = keys.next();
+					if (key != null) {
+						next = Variable.convertIfOldPlayer(subName + key, local, event, Variables.getVariable(subName + key, event, local));
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/variables/Variables.java:515]: unfortunate... we should aim to add an api for variable conversion (or find a way to move this) (not in this pr of course)

File: src/main/java/ch/njol/skript/sections/SecFilter.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+	@UnknownNullability
+	private Variable<?> unfilteredObjects;
+	private final List<Condition> conditions = new ArrayList<>();
+	private boolean isAny;
+
+	@Nullable
+	private Object currentValue;
+	@UnknownNullability
+	private String currentIndex;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/SecFilter.java:71]: move these onto the type? 🙂

File: src/main/java/ch/njol/skript/sections/SecFilter.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+	@Override
+	public @UnknownNullability String getCurrentIndex() {
+		return currentIndex;
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "filter " + unfilteredObjects.toString(event, debug) + " to match " + (isAny ? "any" : "all");
+	}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/SecFilter.java:190]:
```suggestion
	}

```

File: src/main/java/ch/njol/skript/sections/SecFilter.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+		if (sectionNode.isEmpty()) {
+			Skript.error("filter sections must contain at least one condition");
+			return false;
+		}
+		InputSource.InputData inputData = getParser().getData(InputSource.InputData.class);
+		InputSource originalSource = inputData.getSource();
+		inputData.setSource(this);
+		try {
+			for (Node childNode : sectionNode) {
+				if (childNode instanceof SectionNode) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/SecFilter.java:94]: do you think it would be better to check that it's *not* a simple node?

**Review by APickledWalrus — APPROVED** [review]

File: src/main/java/ch/njol/skript/sections/SecFilter.java (hunk 3 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/sections/SecFilter.java:71]: I will do this shortly in a following PR! some refactoring is coming to all these input classes in general

File: src/main/java/ch/njol/skript/sections/SecFilter.java (hunk 5 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/sections/SecFilter.java:94]: yes good idea

**Pull request merged**